monai/transforms/post/array.py

```diff
@@ -183,52 +184,90 @@ class KeepLargestConnectedComponent(Transform):
 
         Returns:
-            A PyTorch Tensor with shape (batch_size, 1, spatial_dim1[, spatial_dim2, ...]).
+            A PyTorch Tensor with shape (batch_size, C, spatial_dim1[, spatial_dim2, ...]).
         """
         channel_dim = 1
         if img.shape[channel_dim] == 1:
+            if self.applied_values is None:
+                raise ValueError("Please provide applied_values for 1 channel data.")
+
             img = torch.squeeze(img, dim=channel_dim)
-        else:
-            raise ValueError("Input data have more than 1 channel.")
 
-        if self.independent:
-            for i in self.applied_values:
-                foreground = (img == i).type(torch.uint8)
+            if self.independent:
+                for i in self.applied_values:
+                    foreground = (img == i).type(torch.uint8)
+                    mask = get_largest_connected_component_mask(foreground, self.connectivity)
+                    img[foreground != mask] = self.background
+            else:
+                foreground = torch.zeros_like(img)
+                for i in self.applied_values:
+                    foreground += (img == i).type(torch.uint8)
                 mask = get_largest_connected_component_mask(foreground, self.connectivity)
                 img[foreground != mask] = self.background
+            output = torch.unsqueeze(img, dim=channel_dim)
         else:
-            foreground = torch.zeros_like(img)
-            for i in self.applied_values:
-                foreground += (img == i).type(torch.uint8)
-            mask = get_largest_connected_component_mask(foreground, self.connectivity)
-            img[foreground != mask] = self.background
-
-        return torch.unsqueeze(img, dim=channel_dim)
+            # one-hot data
+            if self.applied_channel_indices is None:
+                raise ValueError("Please provide applied_channel_indices for one-hot data.")
+
+            # one-hot data is assumed to have binary value in each channel
+            if self.independent:
+                for i in self.applied_channel_indices:
+                    foreground = img[:, i, ...].type(torch.uint8)
+                    mask = get_largest_connected_component_mask(foreground, self.connectivity)
+                    img[:, i, ...][foreground != mask] = 0
+            else:
+                applied_img = img[:, self.applied_channel_indices, ...].type(torch.uint8)
+                foreground = torch.any(applied_img, dim=channel_dim)
+                mask = get_largest_connected_component_mask(foreground, self.connectivity)
+                background_mask = torch.unsqueeze(foreground != mask, dim=channel_dim)
+                background_mask = torch.repeat_interleave(
+                    background_mask, len(self.applied_channel_indices), dim=channel_dim
+                )
+                applied_img[background_mask] = 0
+                img[:, self.applied_channel_indices, ...] = applied_img.type(img.type())
```

Comment:
Is `applied_img` a shallow copy of `img`? Maybe you can test to remove this line 270.

Reply:
OK, sounds good.